src/server/kernel/winputdevice.cpp

```diff
@@ -104,4 +107,39 @@ QInputDevice *WInputDevice::qtDevice() const
     return d->qtDevice;
 }
 
+void WInputDevice::setExclusiveGrabber(QObject *grabber)
+{
+    W_D(WInputDevice);
+    auto pointerDevice = qobject_cast<QPointingDevice*>(d->qtDevice);
+    if (!pointerDevice)
+        return;
+    auto dd = QPointingDevicePrivate::get(pointerDevice);
+    if (dd->activePoints.isEmpty())
+        return;
+    auto firstPoint = dd->activePoints.values().first();
+    dd->setExclusiveGrabber(nullptr, firstPoint.eventPoint, grabber);
+}
+
+QObject *WInputDevice::exclusiveGrabber() const
+{
+    W_DC(WInputDevice);
+    auto pointerDevice = qobject_cast<QPointingDevice*>(d->qtDevice);
+    if (!pointerDevice)
+        return nullptr;
+    auto dd = QPointingDevicePrivate::get(pointerDevice);
+    return dd->firstPointExclusiveGrabber();
+}
+
+QObject *WInputDevice::hoverTarget() const
+{
+    W_DC(WInputDevice);
+    return d->hoverTarget;
+}
+
+void WInputDevice::setHoverTarget(QObject *object)
+{
+    W_D(WInputDevice);
+    d->hoverTarget = object;
```

Comment:
if （object == d->hoverTarget）
   return;

Reply:
Not need, there are only a signal line code